### DISCUSSION TITLE
Chess overlapping on board

### DISCUSSION THREAD

**Discussion by klodha** [rexrainbow/phaser3-rex-notes#136] [Q&A]
Hi,

Thanks for nice plugin, very useful. I am encountering a use case where chess may overlap on board. Basically allowing two chess pieces to have same TileXY. 

Right now when I tried this with monopoly board, its failing and giving following error:
```
Uncaught TypeError: Cannot read property 'x' of null
    at (index):441
    at Array.forEach (<anonymous>)
    at ChessA.moveForward ((index):440)
```

After adding debugger, I see that for that chess object `this.monopoly.chessData.tileXYZ` is null. This happens only when two chess objects are having same tile X,Y. Otherwise `this.monopoly.chessData.tileXYZ` return correct value. 

Please suggest how I can allow two chess pieces on board to have same `this.monopoly.chessData.tileXYZ`.

**Comment by rexrainbow** [comment]
From [these descriptions](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/board/#add-chess)
```
Any (tileX, tileY, tileZ) index contains only 1 chess.
(tileX, tileY) could have more then 1 chess with different tileZ index.
```

Try to set tileZ to different values to avoid overlapping at the same (tileX, tileY, tileZ).

**Reply by klodha** [reply]
```                    
var chessA = new ChessA(board, {
                        x: 1,
                        y: 6,
                        z: 1,
                    }, 0x00ff00, routePathGreen);
                    
                    var chessB = new ChessA(board, {
                        x: 1,
                        y: 6,
                        z: 2,
                    }, 0xffff00, routePathGreen);
```

I am passing different Z values already, but that doesn't seem to be working. I am creating a codepen sample to demonstrate the problem and will share link here.

**Reply by rexrainbow** [reply]
Since this is a bug, it is better to create an issue to report it, thank you.

**Reply by klodha** [reply]
Okay, I will check this once more, just in case if I am doing something wrong. If behavior is confirmed, I will create a bug report. I am also going though board plugin code and trying to debug my issue with non-minified version of plugin. 

So far I see that when AddChess is called, it check occupied (dispatch kickout event, which I am getting). After that its calling removeChess once and then adding it back to the board with updated XYZ. Thats the point when due to same XY value, one of my object fails to add back to boardData, and thus moveTo fails (because board is set null after removeChess call). 

As per code and your comment above, Z should make a difference here. So I am trying with Z property to see if that solves the uid issue (because otherwise its generating same UIDs).

**Reply by rexrainbow** [reply]
[Here](https://codepen.io/rexrainbow/pen/VwKPGPa) is a test case : two chess start at the same tileXY position and move with the same steps. This test case works fine.

**Reply by klodha** [reply]
Thanks, managed to get it working by setting different Z for each chess object. However, with that got another use case. If I use different Z values then Board never emits `kickout` event.  Codepen demonstrating issue is [here](https://codepen.io/klodha83/pen/ExgWYPK?editors=1111).

Is it possible to get `kickout` event but not to remove chess from board even with different Z?

Under board plugin, AddChess.js file I see this block is responsible for that part. 
```
    var occupiedChess = this.tileXYZToChess(tileX, tileY, tileZ);
    if (occupiedChess) {
        this.emit('kickout', gameObject, occupiedChess, curTileXYZ);
    }

    this.removeChess(gameObject);

    if (occupiedChess) {
        this.removeChess(occupiedChess, tileX, tileY, tileZ);
    }

    this.boardData.addUID(this.getChessUID(gameObject), tileX, tileY, tileZ);

    if (this.isBoard) {
        this.getChessData(gameObject).setBoard(this);
    }
```
As I understand, removeChess set `board` property to null on chess and remove it from UID collection. I tried to modify this part to exclude Z check for occupancy and that worked. However issue which I was originally facing popped up back with that, as next it try to add chess back to list, but if Z values are same, then UID is generated as duplicate and thus only 1 chess is now in UID collection making second chess orphan and not linked to board. 

Passing uid during chess creation didn't work as in GetChessData class, its using uidKey property to check and if not found it start from 1 while if this key is found, chess is not added to collection. So in both cases that idea of setting uid before hand didn't work. 

Do you have any suggestion in this case? I know this is different than most of board games, but the game I am trying to simulate with this plugin allow chess of same player to share same tile, but if chess belongs to different player, it is removed (but that also has safe house concepts, meaning special tiles where both player's chess can be at same time).

**Reply by rexrainbow** [reply]
'kickout' event is used for warning user that one chess has been removed unexpectedly, i.e. it is better to prevent 'kickout' in most of case.

**Reply by klodha** [reply]
Thanks. Okay, I can manage kickout check at my game level by comparing the positions of all chess. Its not a big issue. 

Though I think firing `kickout` event can help in most of the cases, as it says `Board will fire kickout event when adding chess to an occupied grid.` So that means target XY already has a chess and a new chess is being allocated same position on the board. This is classic case of most of turn based board games which allow player knock outs.

Thanks for your prompt replies and help with solving the original issue.

**Reply by rexrainbow** [reply]
In my moveTo behavior of Board plugin, it can **pass through** chess which has the same tileZ. The magic is to [change current tileZ to another value temporarily](https://github.com/rexrainbow/phaser3-rex-notes/blob/master/plugins/board/moveto/CanMoveToTile.js#L28-L34), after passing through, [set tileZ back](https://github.com/rexrainbow/phaser3-rex-notes/blob/master/plugins/board/moveto/CanMoveToTile.js#L41-L46).
